apps/src/dropletUtils.js

```diff
@@ -158,52 +158,44 @@ standardConfig.categories = {
  * @returns {Array}
  */
 function mergeFunctionsWithConfig(codeFunctions, dropletConfig, otherConfig) {
+  if (!codeFunctions || !dropletConfig || !dropletConfig.blocks) {
+    return [];
+  }
+
   var merged = [];
 
-  if (codeFunctions && dropletConfig && dropletConfig.blocks) {
-    var blockSets = [ dropletConfig.blocks ];
-    if (otherConfig) {
-      blockSets.splice(0, 0, otherConfig.blocks);
-    }
-    // codeFunctions is an object with named key/value pairs
-    //  key is a block name from dropletBlocks or standardBlocks
-    //  value is an object that can be used to override block defaults
-    for (var s = 0; s < blockSets.length; s++) {
-      var blocks = blockSets[s];
-      for (var i = 0; i < blocks.length; i++) {
-        var block = blocks[i];
-        if (blocks[i].func in codeFunctions) {
-          // We found this particular block, now override the defaults with extend
-          merged.push(utils.extend(blocks[i], codeFunctions[blocks[i].func]));
-        }
+  var blockSets = [ dropletConfig.blocks ];
+  if (otherConfig) {
+    blockSets.splice(0, 0, otherConfig.blocks);
+  }
+
+  // codeFunctions is an object with named key/value pairs
+  //  key is a block name from dropletBlocks or standardBlocks
+  //  value is an object that can be used to override block defaults
+  for (var s = 0; s < blockSets.length; s++) {
+    var set = blockSets[s];
+    for (var i = 0; i < set.length; i++) {
+      var block = set[i];
+      if (block.func in codeFunctions) {
+        // We found this particular block, now override the defaults with extend
+        merged.push($.extend({}, block, codeFunctions[block.func]));
       }
     }
   }
+
   return merged;
 }
 
-//
-// Return a new categories object with the categories from dropletConfig
-// merged with the ones in standardConfig
-//
-
+/**
+ * Return a new categories object with the categories from dropletConfig (app
+ * specific configuration) merged with the ones in standardConfig (global
+ * configuration). App configuration takes precendence
+ */
 function mergeCategoriesWithConfig(dropletConfig) {
-  var merged = {};
-
-  if (dropletConfig && dropletConfig.categories) {
-    var categorySets = [ dropletConfig.categories, standardConfig.categories ];
-    for (var s = 0; s < categorySets.length; s++) {
-      var categories = categorySets[s];
-      for (var catName in categories) {
-        if (!(catName in merged)) {
-          merged[catName] = utils.shallowCopy(categories[catName]);
-        }
-      }
-    }
-  } else {
-    merged = standardConfig.categories;
-  }
-  return merged;
+  // Clone our merged categories so that as we mutate it, we're not mutating
+  // our original config
+  return _.cloneDeep($.extend({}, standardConfig.categories,
+    dropletConfig && dropletConfig.categories));
```

Comment:
The interaction of two issues was causing us to see some test failures
(1) Using $.extend instead of our previous merge technique led to Object.keys being ordered differently than before
(2) Because we were doing a shallowCopy, merged.categories.blocks === standardConfig.categories.blocks. This doesn't change in using $.extend. In generateDropletPalette we end up adding blocks to the returned categories we get from this method. This means we're also adding those blocks to standardConfig. The result is that if I run test1 that calls generateDropletPalette and then I run test2 that calls generateDropletPalette, standardConfig will be (unexpectedly) different.

Consider the case where I have a block in codeFunctions. In a world where standardConfig is not being mutated across runs, in generateDropletPalette I end up generating a set of categories where one category has one block and the rest have zero. As such, I show a single category in the palette, regardless of how the keys are ordered. In the case where standardConfig IS being mutated, I end up having something like 6 categories with blocks (because I have blocks from previous runs), and in order to have my category shown as the default one in the palette, I depend on the order of Object.keys, which has changed because of #1

I think that the only other way that the change in Object.keys ordering could get us into trouble is if we specified in codeFunctions a function whose category is listed in standardConfig but not in the apps dropletConfig. I don't think there's reason this should ever happen, but if it did, it would probably mean that the categories are ordered differently.